### ISSUE TITLE
Programming the  AVR

### ISSUE THREAD

**Issue by TAGood827** [hooyah/SevenSegments#2]
I have built the board, I believe everything is soldered cleanly and in the correct orientations. WE (my boss and I) used Atmel Studio 7 to compile and we have the ATMEGA328P selected for the chip. Everything appears to compile correctly. We can upload to the chip using AVRDUDE setting only the low byte fuses to 0xFF, we see the led flash during upload and we see reports that everything was successful but the chip does not appear to execute the code afterward. There's no blink of the led to indicate it is active but I don't think there is anything in the code that does that. We have the ESP module programmed correctly and we believe that is functioning correctly as we do see that blink about every 5 seconds.

Are we missing setting something during compile or programming phases that is causing the  chip not to execute?

**Comment by hooyah** [comment]
Hey, I'm not sure I can help diagnose your problem remotely, but here are a few pointers.
When the main loop is running the LED on the board should blink around every 10 seconds.
Make sure the fuses are set correctly, so that the crystal is working and the clock is running at 16Mhz.
My fuses are set like this
//	.low = 0xFF,
//	.high = 0xD9,
//	.extended = 0xFF,
Once this works and the board happily blinks away, you can connect a terminal to the serial RX/TX pins on the headers, (eg FTDI board at 5V levels, 9600 baud),  and send commands to it (see code). "$DBG\n" will put it in verbose mode and it will print a whole lot of info to the terminal about what it's doing.
Hope this helps.

**Comment by TAGood827** [comment]
[image: image.png]

**Comment by TAGood827** [comment]
Additional details are as follows:

My boss said he also did something a little off in the compiling with
Studio 7 ("It's been a long time since I used it.") and once he got that
sorted out (he could not tell me what it was as he can't really remember)
things went smoothly from there. He's now back to breeding his pine
trees...😁

We used 0xFF for the low, 0xD9 for the high and had to use 0x05 for the
extended. It would not allow us to do the 0xFF for the extended for some
reason.

It took me a little while to get the gears aligned correctly. It seems like
I still might be off a tooth somewhere yet as during some of the
transitions a segment will be slightly starting to turn for the next digit.
0 to 1 has Segment 7 slightly turning away from "on".
1 to 2 has segment 7 slightly turning back from "off".
2 to 3 looks good
3 to 4 has Segment 6 slightly turning away from "on".
4 to 5 looks good
5 to 6 has segment 7slightly turning back from "off".
6 to 7 looks good
7 to 8 looks good
8 to 9 looks good
9 to 0 looks good

If you have any ideas about what gear to make an adjustment on to correct
this, I would appreciate your input.

Also, my steppers turned in reverse. I had to swap the coil ends for both
coils to get it to turn in the correct direction. I had purchased the
Elegoo 5x combo box from Amazon. They are all marked 28BYJ-48 so no idea
what happened. Unless it has something to do with the Northern
Hemisphere... 🤔

Time to get working on the remaining digits! This is really a marvelous
design and an astounding feat of engineering. Very much the conversation
piece here at work.

Thank you again for your assistance in getting my build going!


Tim

**Comment by TAGood827** [comment]
Today's developments!

We have successfully connected the Rx line and we can see the time being
sent out by the ESP. On Pin 3 on the IN jack, we can see the full time
string, on pin 3 of the OUT jack We see the truncated string.

We were unable to connect the Tx and send in the debug command for some
reason, it doesn't seem to accept any input. We recompiled the code with
the debug mode made active manually and the we could see all the data. The
ADC was not changing much, only from about low 500s to maybe low 600s. I
added a second magnet on top of the first and now we see a very large spike
when it passes by the Hall sensor. It goes from the low 500s to just over
1000 as it passes by. This corresponds with changing from 0 to 1, Is that
correct?

We noticed the DEV would never change, then we spotted in the code that the
trip point was set under 500. We changed it to 995 and now the unit is
calculating a DEV number. It seems to change each time the unit is plugged
in and goes through the initialization routine. Performance hasn't changed
drastically but it is now maintaining an constant offset of 2 on the
display. So when it should be showing 1 it shows 3, etc. I'm not sure what
that is, could I be that far off with my gear setting?

Still trying to perfect the gear alignment for properly timed flipping.

Things that make you go hmmmm...

🤔🤔🤔🤔🤔

**Comment by hooyah** [comment]
The wrong motor rotation is strange, but I have seen many different versions of it, all called the same but with different quirks. Glad you managed to fix the rotation. Maybe I should add a setting for that in the code.
You should see a dip in the numbers when the magnet goes by, at least that's how my unit it set up. It depends on the polarity of the magnet. Since you are getting an offset, try reversing the magnet (and return the threshold to what it was in the code). You may be calculating the wrong peak otherwise.
When adjusting the segments to each other, follow the gear train. The motor is driving segments 6 and 7, match them first to each other. Then match 1 and 2 to 6. At last match 3, 4 and 5 to each other and the rest. After an adjustment it may take several motor turns before you see the results of your changes.
keep this direction in mind:
the motor drives segment 7
7 -> 6
   6 ->1,3,4,5
       1 -> 2
Every adjustment you make to a segment will affect the segments down the line.
It's a bit of a puzzle.

**Comment by TAGood827** [comment]
Okay, flipping the magnet to South (now ranges from 510s to 170s) made a
big difference as the unit now shows the correct number! Did I miss that in
the assembly notes somewhere?

It has also made a difference in the flipping and I am much closer to
showing the digits perfectly.

Thanks again!

Tim

**Comment by TAGood827** [comment]
DOH!

The second magnet was tripping Segment 5 early on the change to 6. So I
pulled that and set the threshold to 475 as we are not getting to 435 and
things are perfect now. The gears are now aligned properly and all digits
are displaying perfectly. Just had to adjust 6 & 7 by one tooth each and
all is well.

Sorry for the confusion!

Thank you again for a truly marvelous design! I will post my build once I
complete the other digits.

Tim

**Comment by TAGood827** [comment]
I found that the relationship of the Segment 5 gear with the magnet makes a
lot of difference in the operation of the digit. I had all the gear marks
aligned as best I could and then as I tried to adjust it seemed like it was
sort of chasing things around the segments by making one segment better and
one segment worse. I reset again and then shifted the 5 gears and it made
the most dramatic difference in the timing. Anything that is off is so very
slight now as to not be noticeable.

I am still struggling with Txing into the unit. Can I do that while the ESP
module is connecting and running? The Tx should be going into Pin 3 of the
IN jack, right? And it is 5V and not TTL, right?

Thanks,

Tim
